patchwork-networking/src/main/java/net/minecraftforge/fml/network/simple/IndexedMessageCodec.java

```diff
@@ -0,0 +1,139 @@
+package net.minecraftforge.fml.network.simple;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
+import it.unimi.dsi.fastutil.shorts.Short2ObjectArrayMap;
+import net.minecraftforge.fml.network.NetworkEvent;
+import net.minecraftforge.fml.network.NetworkInstance;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.MarkerManager;
+
+import net.minecraft.util.PacketByteBuf;
+
+public class IndexedMessageCodec {
+	private static final Logger LOGGER = LogManager.getLogger();
+	private static final Marker SIMPLENET = MarkerManager.getMarker("SIMPLENET");
+	private final Short2ObjectArrayMap<MessageHandler<?>> indicies = new Short2ObjectArrayMap<>();
+	private final Object2ObjectArrayMap<Class<?>, MessageHandler<?>> types = new Object2ObjectArrayMap<>();
+	private final NetworkInstance networkInstance;
+
+	public IndexedMessageCodec() {
+		this(null);
+	}
+
+	public IndexedMessageCodec(final NetworkInstance instance) {
+		this.networkInstance = instance;
+	}
+
+	private static <M> void tryDecode(PacketByteBuf payload, Supplier<NetworkEvent.Context> context, int payloadIndex, MessageHandler<M> codec) {
+		codec.decoder.map(d -> d.apply(payload)).
```

Comment:
I think the functional stuff here is kinda redundant and only obscures the meaning of the function. It would probably be better to just replace the optional with a null check.

Reply:
will do where possible, but i will leave public return values for the sake of accidentally breaking a mod somewhere